receiver/googlecloudspannerreceiver/internal/filter/itemcardinality_test.go

```diff
@@ -0,0 +1,342 @@
+
+	assert.Equal(t, limitByTimestamp, itemFilter.LimitByTimestamp())
+}
+
+func TestItemCardinalityFilter_CanIncludeNewItem(t *testing.T) {
+	logger := zaptest.NewLogger(t)
+	testCases := map[string]struct {
+		totalLimit         int
+		limitByTimestamp   int
+		keysAlreadyInCache []string
+		expectedResult     bool
+	}{
+		"No items in cache and timestamp limit hasn't been exhausted": {2, 1, nil, true},
+		"Cache is full and timestamp limit hasn't been exhausted":     {2, 1, []string{"qwerty1", "qwerty2"}, false},
+		"No items in cache and timestamp limit has been exhausted":    {2, 0, nil, false},
+	}
+
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			filter, err := NewItemCardinalityFilter(metricName, testCase.totalLimit, testCase.limitByTimestamp,
+				itemActivityPeriod, logger)
+			require.NoError(t, err)
+			filterCasted := filter.(*itemCardinalityFilter)
+			defer executeShutdown(t, filterCasted)
+
+			for _, key := range testCase.keysAlreadyInCache {
+				err = filterCasted.cache.Set(key, byte(1))
+				require.NoError(t, err)
+			}
+
+			assert.Equal(t, testCase.expectedResult, filterCasted.canIncludeNewItem(testCase.limitByTimestamp))
+		})
+	}
+}
+
+func TestItemCardinalityFilter_Shutdown(t *testing.T) {
+	logger := zaptest.NewLogger(t)
+	testCases := map[string]struct {
+		closeCache  bool
+		expectError bool
+	}{
+		"Happy path":            {false, false},
+		"Cache has been closed": {true, true},
+	}
+
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			filter, err := NewItemCardinalityFilter(metricName, totalLimit, limitByTimestamp, itemActivityPeriod, logger)
+			require.NoError(t, err)
+			filterCasted := filter.(*itemCardinalityFilter)
+
+			if testCase.closeCache {
+				// Covering case when by some reasons cache is closed
+				err = filterCasted.cache.Close()
+				require.NoError(t, err)
+			}
+
+			if testCase.expectError {
+				require.Error(t, filter.Shutdown())
+			} else {
+				require.NoError(t, filter.Shutdown())
+			}
+		})
+	}
+}
+
+func TestItemCardinalityFilter_Filter(t *testing.T) {
+	items := initialItems(t)
+	logger := zaptest.NewLogger(t)
+	filter, err := NewItemCardinalityFilter(metricName, totalLimit, limitByTimestamp, itemActivityPeriod, logger)
+	require.NoError(t, err)
+	filterCasted := filter.(*itemCardinalityFilter)
+	defer executeShutdown(t, filterCasted)
+
+	filteredItems, err := filter.Filter(items)
+	require.NoError(t, err)
+
+	// Items with key3 and key6 must be not present in filtered items
+	assertInitialFiltering(t, expectedFilteredInitialItems(t), filteredItems)
+
+	items = additionalTestData(t)
+	filteredItems, err = filter.Filter(items)
+	require.NoError(t, err)
+
+	// Cache timeout hasn't been reached, so filtered out all items
+	assert.Equal(t, 0, len(filteredItems))
+
+	// Doing this to avoid of relying on timeouts and sleeps(avoid potential flaky tests)
+	syncChannel := make(chan bool)
```

Comment:
For this specific case, using a `sync.WaitGroup` would probably be more appropriate:

```go
	wg := &sync.WaitGroup{}
	wg.Add(1)
	filterCasted.cache.SetExpirationCallback(func(key string, value interface{}) {
		if filterCasted.cache.Count() > 0 {
			// Waiting until cache is really empty - all items are expired
			return
		}
		wg.Done()
	})
	wg.Wait()
```

Reply:
Previous time it didn't work, as you can remember and led to flaky test - so I accepted this channel approach.  I can try to experiment with this, but will not promise that this will be changed, since current approach is working.

Reply:
Just checked and it doesn't work by the same reason as previous time. With channel works perfect.

Reply:
Just out of curiosity: why is this not working? In that other PR, the problem was that there could have been multiple calls to the callback, so that `wg.Done` was being called twice, which caused a panic. Is this the same here?

Reply:
Looks like yes. I tried to apply proposed changes and ran tests few times and as a result: periodically got the same situation as previously.